Async feature transfer manager

cr0 opened the pull request:
Some plugins like s3 or swift (see #276 or #448) do not ensure that a file is actually written after upload or move operations. This feature allows a plugin to wait after upload and move operations until a file is downloadable. 

Open question: Should we use download to verify if a file exists (could be heavy in some cases) or should we create a feature extension with an `exists` method?

You can try this by annotating the `LocalTransferManager` with `@Async` and add this code to the upload operation to simulate said plugins:
```java
	@Override
	public void upload(final File localFile, final RemoteFile remoteFile) throws StorageException {
		connect();

		final File repoFile = getRemoteFile(remoteFile);
		final File tempRepoFile = new File(getAbsoluteParentDirectory(repoFile) + File.separator + ".temp-" + repoFile.getName());

		// Do not overwrite files with same size!
		if (repoFile.exists() && repoFile.length() == localFile.length()) {
			return;
		}

		// No such local file
		if (!localFile.exists()) {
			throw new StorageException("No such file on local disk: " + localFile);
		}

		new Thread(new Runnable() {
			@Override
			public void run() {
				try {
					logger.log(Level.FINE, "Waiting to store file " + remoteFile);
					Thread.sleep(3000);
					logger.log(Level.FINE, "Storing file now " + remoteFile);
					FileUtils.copyFile(localFile, tempRepoFile);
					FileUtils.moveFile(tempRepoFile, repoFile);
				}
				catch (Exception ex) {
					logger.log(Level.SEVERE, "Unable to copy file " + localFile + " to local repository " + repoFile, ex);
				}
			}
		}).start();
	}
```

sample log
```
8-5-15 20:30:42.010 | RemoteTransacti | main           | INFO : - Uploading /private/tmp/11/.syncany/cache/database-LJUysNFiLSvODRJqLqqX-0000000004 to temp. file RemoteFile[name=temp-SDxif-database-LJUysNFiLSvODRJqLqqX-0000000004] ...
8-5-15 20:30:42.011 | LocalTransferMa | Thread-6       | FINE : Waiting to store file RemoteFile[name=temp-SDxif-database-LJUysNFiLSvODRJqLqqX-0000000004]
8-5-15 20:30:42.011 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 100ms ...
org.syncany.plugins.transfer.StorageFileNotFoundException: No such file in local repository: /private/tmp/11/../22/temporary/temp-SDxif-database-LJUysNFiLSvODRJqLqqX-0000000004
	at org.syncany.plugins.local.LocalTransferManager.download(LocalTransferManager.java:145)
	at org.syncany.plugins.transfer.features.AsyncFeatureTransferManager.waitForFile(AsyncFeatureTransferManager.java:127)
	at org.syncany.plugins.transfer.features.AsyncFeatureTransferManager.upload(AsyncFeatureTransferManager.java:81)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager$6.execute(RetriableFeatureTransferManager.java:124)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.retryMethod(RetriableFeatureTransferManager.java:210)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.upload(RetriableFeatureTransferManager.java:121)
	at org.syncany.plugins.transfer.features.TransactionAwareFeatureTransferManager.upload(TransactionAwareFeatureTransferManager.java:134)
	at org.syncany.plugins.transfer.RemoteTransaction.uploadAndMoveToTempLocation(RemoteTransaction.java:222)
	at org.syncany.plugins.transfer.RemoteTransaction.commit(RemoteTransaction.java:142)
	at org.syncany.plugins.transfer.RemoteTransaction.commit(RemoteTransaction.java:131)
	at org.syncany.operations.up.UpOperation.execute(UpOperation.java:205)
	at org.syncany.cli.UpCommand.execute(UpCommand.java:58)
	at org.syncany.cli.CommandLineClient.runLocally(CommandLineClient.java:392)
	at org.syncany.cli.CommandLineClient.runCommand(CommandLineClient.java:382)
	at org.syncany.cli.CommandLineClient.start(CommandLineClient.java:198)
	at org.syncany.Syncany.main(Syncany.java:52)
8-5-15 20:30:42.113 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 300ms ...
org.syncany.plugins.transfer.StorageFileNotFoundException: No such file in local repository: /private/tmp/11/../22/temporary/temp-SDxif-database-LJUysNFiLSvODRJqLqqX-0000000004
	at org.syncany.plugins.local.LocalTransferManager.download(LocalTransferManager.java:145)
	at org.syncany.plugins.transfer.features.AsyncFeatureTransferManager.waitForFile(AsyncFeatureTransferManager.java:127)
	at org.syncany.plugins.transfer.features.AsyncFeatureTransferManager.upload(AsyncFeatureTransferManager.java:81)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager$6.execute(RetriableFeatureTransferManager.java:124)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.retryMethod(RetriableFeatureTransferManager.java:210)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.upload(RetriableFeatureTransferManager.java:121)
	at org.syncany.plugins.transfer.features.TransactionAwareFeatureTransferManager.upload(TransactionAwareFeatureTransferManager.java:134)
	at org.syncany.plugins.transfer.RemoteTransaction.uploadAndMoveToTempLocation(RemoteTransaction.java:222)
	at org.syncany.plugins.transfer.RemoteTransaction.commit(RemoteTransaction.java:142)
	at org.syncany.plugins.transfer.RemoteTransaction.commit(RemoteTransaction.java:131)
	at org.syncany.operations.up.UpOperation.execute(UpOperation.java:205)
	at org.syncany.cli.UpCommand.execute(UpCommand.java:58)
	at org.syncany.cli.CommandLineClient.runLocally(CommandLineClient.java:392)
	at org.syncany.cli.CommandLineClient.runCommand(CommandLineClient.java:382)
	at org.syncany.cli.CommandLineClient.start(CommandLineClient.java:198)
	at org.syncany.Syncany.main(Syncany.java:52)
8-5-15 20:30:42.419 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 900ms ...
org.syncany.plugins.transfer.StorageFileNotFoundException: No such file in local repository: /private/tmp/11/../22/temporary/temp-SDxif-database-LJUysNFiLSvODRJqLqqX-0000000004
	at org.syncany.plugins.local.LocalTransferManager.download(LocalTransferManager.java:145)
	at org.syncany.plugins.transfer.features.AsyncFeatureTransferManager.waitForFile(AsyncFeatureTransferManager.java:127)
	at org.syncany.plugins.transfer.features.AsyncFeatureTransferManager.upload(AsyncFeatureTransferManager.java:81)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager$6.execute(RetriableFeatureTransferManager.java:124)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.retryMethod(RetriableFeatureTransferManager.java:210)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.upload(RetriableFeatureTransferManager.java:121)
	at org.syncany.plugins.transfer.features.TransactionAwareFeatureTransferManager.upload(TransactionAwareFeatureTransferManager.java:134)
	at org.syncany.plugins.transfer.RemoteTransaction.uploadAndMoveToTempLocation(RemoteTransaction.java:222)
	at org.syncany.plugins.transfer.RemoteTransaction.commit(RemoteTransaction.java:142)
	at org.syncany.plugins.transfer.RemoteTransaction.commit(RemoteTransaction.java:131)
	at org.syncany.operations.up.UpOperation.execute(UpOperation.java:205)
	at org.syncany.cli.UpCommand.execute(UpCommand.java:58)
	at org.syncany.cli.CommandLineClient.runLocally(CommandLineClient.java:392)
	at org.syncany.cli.CommandLineClient.runCommand(CommandLineClient.java:382)
	at org.syncany.cli.CommandLineClient.start(CommandLineClient.java:198)
	at org.syncany.Syncany.main(Syncany.java:52)
8-5-15 20:30:43.323 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 2700ms ...
org.syncany.plugins.transfer.StorageFileNotFoundException: No such file in local repository: /private/tmp/11/../22/temporary/temp-SDxif-database-LJUysNFiLSvODRJqLqqX-0000000004
	at org.syncany.plugins.local.LocalTransferManager.download(LocalTransferManager.java:145)
	at org.syncany.plugins.transfer.features.AsyncFeatureTransferManager.waitForFile(AsyncFeatureTransferManager.java:127)
	at org.syncany.plugins.transfer.features.AsyncFeatureTransferManager.upload(AsyncFeatureTransferManager.java:81)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager$6.execute(RetriableFeatureTransferManager.java:124)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.retryMethod(RetriableFeatureTransferManager.java:210)
	at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.upload(RetriableFeatureTransferManager.java:121)
	at org.syncany.plugins.transfer.features.TransactionAwareFeatureTransferManager.upload(TransactionAwareFeatureTransferManager.java:134)
	at org.syncany.plugins.transfer.RemoteTransaction.uploadAndMoveToTempLocation(RemoteTransaction.java:222)
	at org.syncany.plugins.transfer.RemoteTransaction.commit(RemoteTransaction.java:142)
	at org.syncany.plugins.transfer.RemoteTransaction.commit(RemoteTransaction.java:131)
	at org.syncany.operations.up.UpOperation.execute(UpOperation.java:205)
	at org.syncany.cli.UpCommand.execute(UpCommand.java:58)
	at org.syncany.cli.CommandLineClient.runLocally(CommandLineClient.java:392)
	at org.syncany.cli.CommandLineClient.runCommand(CommandLineClient.java:382)
	at org.syncany.cli.CommandLineClient.start(CommandLineClient.java:198)
	at org.syncany.Syncany.main(Syncany.java:52)
8-5-15 20:30:45.014 | LocalTransferMa | Thread-6       | FINE : Storing file now RemoteFile[name=temp-SDxif-database-LJUysNFiLSvODRJqLqqX-0000000004]
8-5-15 20:30:46.030 | AsyncFeatureTra | main           | FINE : File found RemoteFile[name=temp-SDxif-database-LJUysNFiLSvODRJqLqqX-0000000004]
8-5-15 20:30:46.031 | RemoteTransacti | main           | INFO : - Moving temp. file RemoteFile[name=temp-EwNwm-multichunk-bf24c39b2b9b16bccbeb6070d548eecfd9de039e] to final location RemoteFile[name=multichunk-bf24c39b2b9b16bccbeb6070d548eecfd9de039e] ...

```

pimotte commented:
Downloading here is very expensive. It basically doubles the bandwith usage, so we really need to use exists(). We could also just add the exists() to the Async feature? (Or was that your intention?)

cr0 commented:
You right pim, adding an `exist` method seems to be the right way.

I don't like the idea of pre-waiting when downloading since a file is downloaded much more often than it is uploaded (which should happen only once). And the upload/move operations are actually the ones which cause trouble. So it seems like post-waiting is a better choice (however, I'm open for ideas).

pimotte commented:
Mmmm. Yes, I can get on board with the post-waiting.

binwiederhier commented:
That. Is. Awesome. 
Agree with the post waiting. Love the fact that it's just a feature again. :+1: :+1:

cr0 commented:
There's some error related to an action file. Seems like it is deleted too soon. I think @pimotte introduced action files, right? Could you have a look?

LocalTransferManager to test: https://gist.github.com/cr0/8de92a84ad78c596f977

```
$ sydev up -d
debug
Application version: 0.4.5-alpha+SNAPSHOT.1505090940.gitaa5ea11
9-5-15 11:41:12.623 | CommandLineClie | main           | INFO : Application version: 0.4.5-alpha+SNAPSHOT.1505090940.gitaa5ea11
9-5-15 11:41:12.703 | PidFileUtil     | main           | INFO : isProcessRunningUnixLike(70493) returned 0, process running = true
9-5-15 11:41:12.706 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Registering StatusCommand (org.syncany.cli.StatusCommand@231f98ef) ...
9-5-15 11:41:12.714 | org.syncany.plu | main           | FINE : Validating required fields
9-5-15 11:41:12.717 | TransferManager | main           | INFO : Building LocalTransferManager from config '11' ...
9-5-15 11:41:12.718 | TransferManager | main           | INFO : - With feature Retriable
9-5-15 11:41:12.719 | TransferManager | main           | FINE : - Checking if selected features supported ...
9-5-15 11:41:12.722 | TransferManager | main           | FINE : - Checking for duplicate features ...
9-5-15 11:41:12.724 | TransferManager | main           | FINE : - Checking required feature TransactionAware in LocalTransferManager ...
9-5-15 11:41:12.730 | TransferManager | main           | FINE : - Checking required feature Retriable in LocalTransferManager ...
9-5-15 11:41:12.733 | TransferManager | main           | FINE : - Wrapping TransferManager LocalTransferManager in RetriableFeatureTransferManager
9-5-15 11:41:12.734 | org.syncany.plu | main           | FINE : Validating required fields
9-5-15 11:41:12.735 | TransferManager | main           | INFO : Building LocalTransferManager from config '11' ...
9-5-15 11:41:12.735 | TransferManager | main           | INFO : - With feature Async
9-5-15 11:41:12.735 | TransferManager | main           | INFO : - With feature Retriable
9-5-15 11:41:12.736 | TransferManager | main           | INFO : - With feature PathAware
9-5-15 11:41:12.736 | TransferManager | main           | INFO : - With feature TransactionAware
9-5-15 11:41:12.739 | TransferManager | main           | FINE : - Checking if selected features supported ...
9-5-15 11:41:12.741 | TransferManager | main           | FINE : - Checking for duplicate features ...
9-5-15 11:41:12.741 | TransferManager | main           | FINE : - Checking required feature TransactionAware in LocalTransferManager ...
9-5-15 11:41:12.741 | TransferManager | main           | FINE : - Checking required feature Retriable in LocalTransferManager ...
9-5-15 11:41:12.743 | TransferManager | main           | FINE : - Wrapping TransferManager LocalTransferManager in AsyncFeatureTransferManager
9-5-15 11:41:12.744 | TransferManager | main           | FINE : - Wrapping TransferManager AsyncFeatureTransferManager in RetriableFeatureTransferManager
9-5-15 11:41:12.744 | TransferManager | main           | INFO : - SKIPPING unsupported optional feature PathAware
9-5-15 11:41:12.744 | TransferManager | main           | FINE : - Wrapping TransferManager RetriableFeatureTransferManager in TransactionAwareFeatureTransferManager
9-5-15 11:41:12.753 | DatabaseConnect | main           | INFO : Loading database driver org.hsqldb.jdbcDriver ...
9-5-15 11:41:13.193 | hsqldb.db.HSQLD | main           | INFO : dataFileCache open start
9-5-15 11:41:13.200 | hsqldb.db.HSQLD | main           | INFO : dataFileCache open end
9-5-15 11:41:13.259 | hsqldb.db.HSQLD | main           | INFO : checkpointClose start
9-5-15 11:41:13.265 | hsqldb.db.HSQLD | main           | INFO : dataFileCache commit start
9-5-15 11:41:13.268 | hsqldb.db.HSQLD | main           | FINE : copyShadow [size, time] 81980 1
9-5-15 11:41:13.269 | hsqldb.db.HSQLD | main           | FINE : setFileModified flag set 
9-5-15 11:41:13.271 | hsqldb.db.HSQLD | main           | FINE : cache save rows [count,time] totals 155,4 operation 155,4 txts 176
9-5-15 11:41:13.272 | hsqldb.db.HSQLD | main           | FINE : file sync end
9-5-15 11:41:13.273 | hsqldb.db.HSQLD | main           | INFO : dataFileCache commit end
9-5-15 11:41:13.279 | hsqldb.db.HSQLD | main           | INFO : checkpointClose end
9-5-15 11:41:13.302 | UpOperation     | main           | INFO : 
9-5-15 11:41:13.303 | UpOperation     | main           | INFO : Running 'Sync up' at client LJUysNFiLSvODRJqLqqX ...
9-5-15 11:41:13.303 | UpOperation     | main           | INFO : --------------------------------------------
9-5-15 11:41:13.303 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event UpStartSyncExternalEvent (org.syncany.operations.daemon.messages.UpStartSyncExternalEvent@59a008ba) ...
Starting indexing and upload ...9-5-15 11:41:13.312 | StatusOperation | main           | INFO : 
9-5-15 11:41:13.313 | StatusOperation | main           | INFO : Running 'Status' at client LJUysNFiLSvODRJqLqqX ...
9-5-15 11:41:13.313 | StatusOperation | main           | INFO : --------------------------------------------
9-5-15 11:41:13.313 | StatusOperation | main           | INFO : Querying current file tree from database ...
9-5-15 11:41:13.314 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event StatusStartSyncExternalEvent (org.syncany.operations.daemon.messages.StatusStartSyncExternalEvent@67d18ed7) ...
Checking for new or altered files ...9-5-15 11:41:13.349 | StatusOperation | main           | INFO : Analyzing local folder /private/tmp/11 ...
9-5-15 11:41:13.360 | StatusOperation | main           | FINE : - Ignoring file (syncany app-related): .syncany
9-5-15 11:41:13.362 | StatusOperation | main           | FINE : - New file: fff
9-5-15 11:41:13.365 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event StatusEndSyncExternalEvent (org.syncany.operations.daemon.messages.StatusEndSyncExternalEvent@10a9d961) ...
9-5-15 11:41:13.367 | AbstractTransfe | main           | INFO : Looking for other running remote operations ...
9-5-15 11:41:13.373 | LsRemoteOperati | main           | INFO : 
9-5-15 11:41:13.373 | LsRemoteOperati | main           | INFO : Running 'Remote Status' at client LJUysNFiLSvODRJqLqqX ...
9-5-15 11:41:13.373 | LsRemoteOperati | main           | INFO : --------------------------------------------
9-5-15 11:41:13.374 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event LsRemoteStartSyncExternalEvent (org.syncany.operations.daemon.messages.LsRemoteStartSyncExternalEvent@61fe30) ...
Checking remote changes ...9-5-15 11:41:13.376 | LsRemoteOperati | main           | INFO : Retrieving remote database list.
9-5-15 11:41:13.377 | LsRemoteOperati | main           | INFO : - Remote database database-LJUysNFiLSvODRJqLqqX-0000000001 is already known (in local database). Ignoring.
9-5-15 11:41:13.377 | LsRemoteOperati | main           | INFO : - Remote database database-LJUysNFiLSvODRJqLqqX-0000000002 is already known (in local database). Ignoring.
9-5-15 11:41:13.378 | LsRemoteOperati | main           | INFO : - Remote database database-LJUysNFiLSvODRJqLqqX-0000000003 is already known (in local database). Ignoring.
9-5-15 11:41:13.378 | LsRemoteOperati | main           | INFO : - Remote database database-LJUysNFiLSvODRJqLqqX-0000000004 is already known (in local database). Ignoring.
9-5-15 11:41:13.378 | LsRemoteOperati | main           | INFO : - Remote database database-LJUysNFiLSvODRJqLqqX-0000000005 is already known (in local database). Ignoring.
9-5-15 11:41:13.380 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event LsRemoteEndSyncExternalEvent (org.syncany.operations.daemon.messages.LsRemoteEndSyncExternalEvent@a3d9978) ...
9-5-15 11:41:13.381 | UpOperation     | main           | INFO : No remote changes, ready to upload.
9-5-15 11:41:13.382 | ActionFileHandl | main           | INFO : Starting action for RemoteFile[name=action-up-LJUysNFiLSvODRJqLqqX-1431164472749] ...
9-5-15 11:41:13.382 | ActionFileHandl | main           | INFO : Uploading action file: RemoteFile[name=action-up-LJUysNFiLSvODRJqLqqX-1431164472749]
9-5-15 11:41:13.385 | LocalTransferMa | Thread-3       | FINE : Waiting to store file RemoteFile[name=action-up-LJUysNFiLSvODRJqLqqX-1431164472749]
9-5-15 11:41:13.385 | ActionFileHandl | main           | INFO : Scheduling action renewal task for every 2 minutes, for RemoteFile[name=action-up-LJUysNFiLSvODRJqLqqX-1431164472749] ...
9-5-15 11:41:13.386 | TransactionAwar | main           | INFO : Done rolling back previous transactions.
9-5-15 11:41:13.400 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event UpIndexStartSyncExternalEvent (org.syncany.operations.daemon.messages.UpIndexStartSyncExternalEvent@598bd2ba) ...
Indexing 1 new or altered file(s)...9-5-15 11:41:13.401 | Indexer         | main           | FINE : - +File /private/tmp/11/fff
9-5-15 11:41:13.402 | Indexer         | main           | FINE : - +FileContent: /private/tmp/11/fff
9-5-15 11:41:13.448 | Indexer         | main           | FINE : - Chunk new: b773831c59d405f31754d0175428556c88fea364
9-5-15 11:41:13.455 | Indexer         | main           | FINE : - +MultiChunk 5f64e6e20e3b9a700d55780313a32aa7b68c03fd
9-5-15 11:41:13.498 | Indexer         | main           | FINE : - Chunk > MultiChunk: b773831c59d405f31754d0175428556c88fea364 > 5f64e6e20e3b9a700d55780313a32aa7b68c03fd
9-5-15 11:41:13.499 | Indexer         | main           | FINE : - Chunk > FileContent: b773831c59d405f31754d0175428556c88fea364 > /private/tmp/11/fff
9-5-15 11:41:13.506 | Indexer         | main           | FINE : - Chunk new: 9e9da55d02736baffbabd804da3de8733dc37c8c
9-5-15 11:41:13.539 | Indexer         | main           | FINE : - Chunk > MultiChunk: 9e9da55d02736baffbabd804da3de8733dc37c8c > 5f64e6e20e3b9a700d55780313a32aa7b68c03fd
9-5-15 11:41:13.540 | Indexer         | main           | FINE : - Chunk > FileContent: 9e9da55d02736baffbabd804da3de8733dc37c8c > /private/tmp/11/fff
9-5-15 11:41:13.549 | Indexer         | main           | FINE : - Chunk new: 3d2ff35660d4e74c514e31f4a208679f3f2d76f5
9-5-15 11:41:13.597 | Indexer         | main           | FINE : - Chunk > MultiChunk: 3d2ff35660d4e74c514e31f4a208679f3f2d76f5 > 5f64e6e20e3b9a700d55780313a32aa7b68c03fd
9-5-15 11:41:13.598 | Indexer         | main           | FINE : - Chunk > FileContent: 3d2ff35660d4e74c514e31f4a208679f3f2d76f5 > /private/tmp/11/fff
9-5-15 11:41:13.603 | Indexer         | main           | FINE : - Chunk new: c26397c4262c422046ecee35ba625b22ba4860c6
9-5-15 11:41:13.645 | Indexer         | main           | FINE : - Chunk > MultiChunk: c26397c4262c422046ecee35ba625b22ba4860c6 > 5f64e6e20e3b9a700d55780313a32aa7b68c03fd
9-5-15 11:41:13.645 | Indexer         | main           | FINE : - Chunk > FileContent: c26397c4262c422046ecee35ba625b22ba4860c6 > /private/tmp/11/fff
9-5-15 11:41:13.648 | Indexer         | main           | FINE : - Chunk new: dcac5b4d959f11c0d7f3b48ba81f648bc449ed07
9-5-15 11:41:13.663 | Indexer         | main           | FINE : - Chunk > MultiChunk: dcac5b4d959f11c0d7f3b48ba81f648bc449ed07 > 5f64e6e20e3b9a700d55780313a32aa7b68c03fd
9-5-15 11:41:13.664 | Indexer         | main           | FINE : - Chunk > FileContent: dcac5b4d959f11c0d7f3b48ba81f648bc449ed07 > /private/tmp/11/fff
9-5-15 11:41:13.665 | Indexer         | main           | FINE : - /File: fff (checksum 2a85f682750eab9ec4ad0925c3445384ed59272e)
9-5-15 11:41:13.665 | Indexer         | main           | FINE :    * No old file history found, starting new history (path: fff, checksum: 2a85f682750eab9ec4ad0925c3445384ed59272e)
9-5-15 11:41:13.666 | FileVersionComp | main           | INFO :      - [DELETED]: Local file DIFFERS from file version, actual file is NULL, for file fff
9-5-15 11:41:13.667 | Indexer         | main           | INFO :    * Added file version:    FileVersion [version=1, path=fff, type=FILE, status=NEW, size=2345472, lastModified=Sat May 09 11:41:08 CEST 2015, linkTarget=null, checksum=2a85f682750eab9ec4ad0925c3445384ed59272e, updated=Sat May 09 11:41:13 CEST 2015, posixPermissions=rw-r--r--, dosAttributes=--a-]
9-5-15 11:41:13.667 | Indexer         | main           | INFO :      based on file version: null
9-5-15 11:41:13.667 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event UpIndexChangesDetectedSyncExternalEvent (org.syncany.operations.daemon.messages.UpIndexChangesDetectedSyncExternalEvent@41e68d87) ...
9-5-15 11:41:13.672 | Indexer         | main           | FINE : - /MultiChunk 5f64e6e20e3b9a700d55780313a32aa7b68c03fd
9-5-15 11:41:13.674 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event UpIndexEndSyncExternalEvent (org.syncany.operations.daemon.messages.UpIndexEndSyncExternalEvent@58be6e8) ...
9-5-15 11:41:13.674 | Indexer         | main           | FINE : - Looking for deleted files ...
9-5-15 11:41:13.675 | Indexer         | main           | FINE :   + Deleted: Adding DELETED version: FileVersion [version=2, path=ddd, type=FILE, status=DELETED, size=2502144, lastModified=Fri May 08 20:30:23 CEST 2015, linkTarget=null, checksum=8179616fa3cc39ceaa4f2640b109cb241841e628, updated=Sat May 09 11:41:13 CEST 2015, posixPermissions=rw-r--r--, dosAttributes=--a-]
9-5-15 11:41:13.675 | Indexer         | main           | FINE :                            based on: FileVersion [version=1, path=ddd, type=FILE, status=NEW, size=2502144, lastModified=Fri May 08 20:30:23 CEST 2015, linkTarget=null, checksum=8179616fa3cc39ceaa4f2640b109cb241841e628, updated=Fri May 08 20:30:33 CEST 2015, posixPermissions=rw-r--r--, dosAttributes=--a-]
9-5-15 11:41:13.676 | Indexer         | main           | FINE :   + Deleted: Adding DELETED version: FileVersion [version=2, path=eee, type=FILE, status=DELETED, size=3480576, lastModified=Fri May 08 22:34:33 CEST 2015, linkTarget=null, checksum=54f98d54581601775d9fae68eaaa324eea166eec, updated=Sat May 09 11:41:13 CEST 2015, posixPermissions=rw-r--r--, dosAttributes=--a-]
9-5-15 11:41:13.676 | Indexer         | main           | FINE :                            based on: FileVersion [version=1, path=eee, type=FILE, status=NEW, size=3480576, lastModified=Fri May 08 22:34:33 CEST 2015, linkTarget=null, checksum=54f98d54581601775d9fae68eaaa324eea166eec, updated=Fri May 08 22:34:41 CEST 2015, posixPermissions=rw-r--r--, dosAttributes=--a-]
9-5-15 11:41:13.692 | UpOperation     | main           | INFO : Last vector clock was: (LJUysNFiLSvODRJqLqqX5)
9-5-15 11:41:13.698 | UpOperation     | main           | INFO : Uploading new multichunks ...
9-5-15 11:41:13.706 | UpOperation     | main           | INFO : - Uploading multichunk 5f64e6e20e3b9a700d55780313a32aa7b68c03fd from /private/tmp/11/.syncany/cache/multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd to RemoteFile[name=multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd] ...
9-5-15 11:41:13.710 | RemoteTransacti | main           | INFO : - Adding file to TX for UPLOAD: /private/tmp/11/.syncany/cache/multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd -> Temp. remote file: RemoteFile[name=temp-qPNwQ-multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd], final location: RemoteFile[name=multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd]
9-5-15 11:41:13.714 | UpOperation     | main           | INFO : No DIRTY data found in database (no dirty databases); Nothing to do here.
9-5-15 11:41:13.717 | UpOperation     | main           | INFO : Saving local delta database, version LJUysNFiLSvODRJqLqqX/(LJUysNFiLSvODRJqLqqX6)/T=1431164473697 to file /private/tmp/11/.syncany/cache/database-LJUysNFiLSvODRJqLqqX-0000000006 ... 
9-5-15 11:41:13.717 | UpOperation     | main           | INFO : - Saving database to /private/tmp/11/.syncany/cache/database-LJUysNFiLSvODRJqLqqX-0000000006 ...
9-5-15 11:41:13.729 | UpOperation     | main           | INFO : - Uploading local delta database file ...
9-5-15 11:41:13.729 | UpOperation     | main           | INFO : - Uploading /private/tmp/11/.syncany/cache/database-LJUysNFiLSvODRJqLqqX-0000000006 to RemoteFile[name=database-LJUysNFiLSvODRJqLqqX-0000000006] ...
9-5-15 11:41:13.730 | RemoteTransacti | main           | INFO : - Adding file to TX for UPLOAD: /private/tmp/11/.syncany/cache/database-LJUysNFiLSvODRJqLqqX-0000000006 -> Temp. remote file: RemoteFile[name=temp-mYFmI-database-LJUysNFiLSvODRJqLqqX-0000000006], final location: RemoteFile[name=database-LJUysNFiLSvODRJqLqqX-0000000006]
9-5-15 11:41:13.734 | UpOperation     | main           | INFO : Adding shutdown hook (to allow resuming the upload) ...
9-5-15 11:41:13.734 | RemoteTransacti | main           | INFO : Starting TX.commit() ...
9-5-15 11:41:13.775 | RemoteTransacti | main           | INFO : Wrote transaction manifest to temporary file: /private/tmp/11/.syncany/cache/temp-transaction-1781770767119974313.tmp
9-5-15 11:41:13.776 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event UpUploadFileSyncExternalEvent (org.syncany.operations.daemon.messages.UpUploadFileSyncExternalEvent@63787180) ...
Uploading transaction-5443d039 ...9-5-15 11:41:13.777 | RemoteTransacti | main           | INFO : - Uploading remote transaction file RemoteFile[name=transaction-5443d039] ...
9-5-15 11:41:13.777 | LocalTransferMa | Thread-4       | FINE : Waiting to store file RemoteFile[name=transaction-5443d039]
9-5-15 11:41:13.778 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 100ms ...
9-5-15 11:41:13.883 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 300ms ...
9-5-15 11:41:14.187 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 900ms ...
9-5-15 11:41:15.092 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 2700ms ...
9-5-15 11:41:16.391 | LocalTransferMa | Thread-3       | FINE : Storing file now RemoteFile[name=action-up-LJUysNFiLSvODRJqLqqX-1431164472749]
9-5-15 11:41:16.398 | LocalTransferMa | Thread-3       | SEVE : Unable to copy file /var/folders/7q/9ftg5_q54g9c8tr76gt9hqd40000gn/T/syncany-action-6189741870278546210.tmp to local repository /private/tmp/11/../22/actions/action-up-LJUysNFiLSvODRJqLqqX-1431164472749
java.io.FileNotFoundException: Source '/var/folders/7q/9ftg5_q54g9c8tr76gt9hqd40000gn/T/syncany-action-6189741870278546210.tmp' does not exist
	at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:1074)
	at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:1038)
	at org.syncany.plugins.local.LocalTransferManager$1.run(LocalTransferManager.java:208)
	at java.lang.Thread.run(Thread.java:745)
9-5-15 11:41:16.778 | LocalTransferMa | Thread-4       | FINE : Storing file now RemoteFile[name=transaction-5443d039]
9-5-15 11:41:17.797 | AsyncFeatureTra | main           | FINE : RemoteFile[name=transaction-5443d039] exists on the remote side
9-5-15 11:41:17.798 | RemoteTransacti | main           | INFO : - Starting to upload data in commit.
9-5-15 11:41:17.799 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event UpUploadFileInTransactionSyncExternalEvent (org.syncany.operations.daemon.messages.UpUploadFileInTransactionSyncExternalEvent@2e1ef60) ...
Uploading 1/2 (2.2 MB, total 0%) ...9-5-15 11:41:17.800 | RemoteTransacti | main           | INFO : - Uploading /private/tmp/11/.syncany/cache/multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd to temp. file RemoteFile[name=temp-qPNwQ-multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd] ...
9-5-15 11:41:17.802 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 100ms ...
9-5-15 11:41:17.802 | LocalTransferMa | Thread-5       | FINE : Waiting to store file RemoteFile[name=temp-qPNwQ-multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd]
9-5-15 11:41:17.905 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 300ms ...
9-5-15 11:41:18.206 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 900ms ...
9-5-15 11:41:19.108 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 2700ms ...
9-5-15 11:41:20.803 | LocalTransferMa | Thread-5       | FINE : Storing file now RemoteFile[name=temp-qPNwQ-multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd]
9-5-15 11:41:21.812 | AsyncFeatureTra | main           | FINE : RemoteFile[name=temp-qPNwQ-multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd] exists on the remote side
9-5-15 11:41:21.813 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event UpUploadFileInTransactionSyncExternalEvent (org.syncany.operations.daemon.messages.UpUploadFileInTransactionSyncExternalEvent@22356acd) ...
Uploading 2/2 (850 B, total 100%) ...9-5-15 11:41:21.814 | RemoteTransacti | main           | INFO : - Uploading /private/tmp/11/.syncany/cache/database-LJUysNFiLSvODRJqLqqX-0000000006 to temp. file RemoteFile[name=temp-mYFmI-database-LJUysNFiLSvODRJqLqqX-0000000006] ...
9-5-15 11:41:21.814 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 100ms ...
9-5-15 11:41:21.816 | LocalTransferMa | Thread-6       | FINE : Waiting to store file RemoteFile[name=temp-mYFmI-database-LJUysNFiLSvODRJqLqqX-0000000006]
9-5-15 11:41:21.916 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 300ms ...
9-5-15 11:41:22.218 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 900ms ...
9-5-15 11:41:23.122 | AsyncFeatureTra | main           | FINE : File not found on the remote side, perhaps its in transit, waiting 2700ms ...
9-5-15 11:41:24.817 | LocalTransferMa | Thread-6       | FINE : Storing file now RemoteFile[name=temp-mYFmI-database-LJUysNFiLSvODRJqLqqX-0000000006]
9-5-15 11:41:25.826 | AsyncFeatureTra | main           | FINE : RemoteFile[name=temp-mYFmI-database-LJUysNFiLSvODRJqLqqX-0000000006] exists on the remote side
9-5-15 11:41:25.826 | RemoteTransacti | main           | INFO : - Moving temp. file RemoteFile[name=temp-qPNwQ-multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd] to final location RemoteFile[name=multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd] ...
9-5-15 11:41:25.827 | AsyncFeatureTra | main           | FINE : RemoteFile[name=multichunk-5f64e6e20e3b9a700d55780313a32aa7b68c03fd] exists on the remote side
9-5-15 11:41:25.827 | RemoteTransacti | main           | INFO : - Moving temp. file RemoteFile[name=temp-mYFmI-database-LJUysNFiLSvODRJqLqqX-0000000006] to final location RemoteFile[name=database-LJUysNFiLSvODRJqLqqX-0000000006] ...
9-5-15 11:41:25.828 | AsyncFeatureTra | main           | FINE : RemoteFile[name=database-LJUysNFiLSvODRJqLqqX-0000000006] exists on the remote side
9-5-15 11:41:25.828 | RemoteTransacti | main           | INFO : - Deleting remote transaction file RemoteFile[name=transaction-5443d039] ...
9-5-15 11:41:25.829 | RemoteTransacti | main           | INFO : END of TX.commmit(): Succesfully committed transaction.
9-5-15 11:41:25.829 | RemoteTransacti | main           | INFO : END of TX.delTemp(): Sucessfully deleted final files.
9-5-15 11:41:25.831 | UpOperation     | main           | INFO : Persisting local SQL database (new database version LJUysNFiLSvODRJqLqqX/(LJUysNFiLSvODRJqLqqX6)/T=1431164473697) ...
9-5-15 11:41:25.856 | UpOperation     | main           | INFO : Removing DIRTY database versions from database ...
9-5-15 11:41:25.880 | ActionFileHandl | main           | INFO : Finishing action for RemoteFile[name=action-up-LJUysNFiLSvODRJqLqqX-1431164472749] ...
9-5-15 11:41:25.880 | ActionFileHandl | main           | INFO : Deleting action file: RemoteFile[name=action-up-LJUysNFiLSvODRJqLqqX-1431164472749]
9-5-15 11:41:25.881 | AbstractTransfe | main           | INFO : Cleaning own old action files ...
9-5-15 11:41:25.884 | Cache           | main           | INFO : Cache size okay (90768 KB), no need to clean (keep size is 500 MB)
9-5-15 11:41:25.884 | UpOperation     | main           | INFO : Sync up done.
9-5-15 11:41:25.886 | InternalEventBu | main           | INFO : Event bus 'LocalEventBus': Posting event UpEndSyncExternalEvent (org.syncany.operations.daemon.messages.UpEndSyncExternalEvent@4b3ed2f0) ...
A fff                                
D ddd
D eee
Sync up finished.
```

gtrefs commented on syncany-lib/src/main/java/org/syncany/plugins/transfer/features/AsyncFeatureTransferManager.java:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	private void waitForFile(RemoteFile remoteFile) throws StorageException {
+		while (true) {
```
Maybe you can use a `ListenableFuture` here? At least you would get rid of the busy wating. However, I think to make this happen, the `exists` method needs to be blocking. Here is a sketch:
```java
ListeningExecutorService service = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(10));
ListenableFuture<?> remoteFileExists = service.submit(new Callable<RemoteFile>() {
  public RemoteFile call() {
   if(Thread.isInterrupted()) throw Throwables.throw(new InterruptedException("Interrupt"));      
   asyncFeatureExtension.exists(remoteFile);
   return remoteFile; 
  }
});
Futures.addCallback(explosion, new FutureCallback<?>() {
  public void onSuccess(RemoteFile remoteFile) {
    // ...
  }
  public void onFailure(Throwable thrown) {
    //...
  }
});
remoteFileExists.get(1, SECOND);
```

binwiederhier replied:
Just so that it doesn't get lost: @cr0 responded to me on IRC thinking that I commented on this.

```
[12:09:05] <cr0> i've seen that comment but since the listenable future relies on a blocking exists method (which is managed by plugin devs), i hesitate to change that
[12:09:47] <cr0> no it doesn't because it does not retry anything, it only blocks for few seconds and then let the retrieabletm take over
[12:10:41] <cr0> but good point, the exception's message is wrong
```

binwiederhier commented:
@cr0 What do you think of `@WaitFor` instead. It's not as sexy, but it makes more sense.

binwiederhier commented:
@cr0 We discussed it again on IRC and we (@pimotte, @gtrefs) think that `@ReadWriteConsistent` best describes the feature. If you're okay with it too, can you rename it?

cr0 commented:
Isn't that the opposite? This makes me believe that the plugin is consistent regarding read and write, i.e. a file is immediately available after it is created. However, s3 and swift do not guarantee that so they are inconsistent? Thus imho it should be `@ ReadWriteInconsistent`? All other plugins are consistent by design and they don't have that annotation. I find it confusing.

binwiederhier commented:
Apologies, I meant `@ReadAfterWriteConsistent`. The idea was that this is a plugin specific property, not a property of the backend. A `@PathAware` feature fixes the inability of the backend to deal with large file lists, and a `@TransactionAware` feature "fixes" the TX-unawareness of the backend. 

Does that make sense?

binwiederhier commented:
Reference:
http://shlomoswidler.com/2009/12/read-after-write-consistency-in-amazon.html

cr0 commented:
Ok now I understand your point of view and I think it makes sense.

binwiederhier commented:
nicely put "I think it makes sense", hehe :+1:

cr0 commented:
Ok bug is fixed: the TM for ActionFiles also needs the `ReadAfterWriteConsistent` feature. Please merge if you want.

```
9-5-15 18:39:57.189 | LocalTransferMa | Thread-3       | SEVE : Unable to copy file /var/folders/7q/9ftg5_q54g9c8tr76gt9hqd40000gn/T/syncany-action-6349484014147268567.tmp to local repository /tmp/22/actions/action-up-PpKBqcPIQSinlZlSTQuE-1431189596522
java.io.FileNotFoundException: Source '/var/folders/7q/9ftg5_q54g9c8tr76gt9hqd40000gn/T/syncany-action-6349484014147268567.tmp' does not exist
	at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:1074)
	at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:1038)
	at org.syncany.plugins.local.LocalTransferManager$1.run(LocalTransferManager.java:208)
	at java.lang.Thread.run(Thread.java:745)
```

binwiederhier commented:
Well then, I'll do that :+1:

binwiederhier commented:
Done. Nicely done.
